File: Python/Doubly Linked List.py
class Node:
    def __init__(self, data):
       self.data, self.prev, self.next = data, None, None
 
class DoublyLinkedList:
    def __init__(self):
        self.head, self.tail, self.nodes = None, None, 0
    
    def insertStart(self, data):
        new = Node(data)
        if self.head is None:
            self.head, self.tail = new, new
        else:
            new.next = self.head
            self.head.prev = new
            self.head = new
        self.nodes += 1
    
    def insertMid(self, pos, data):
        if self.head is None:
            new = Node(data)
            self.head, self.tail = new, new
            self.nodes += 1
        else:
            if pos == 1:
                self.insertStart(data)
            elif pos >= self.nodes:
                self.insertEnd(data)
            else:
                new = Node(data)
                curr, temp = self.head, self.head.next
                for _ in range(pos-2):
                    curr, temp = curr.next, temp.next
                curr.next = new
                new.prev = curr
                new.next = temp
                temp.prev = new
                self.nodes += 1
    
    def insertEnd(self, data):
        new = Node(data)
        if self.head is None:
            self.head, self.tail = new, new
        else:
            self.tail.next = new
            new.prev = self.tail
            self.tail = new
        self.nodes += 1
    
    def deleteStart(self):
        if self.head is None:
            return None
        else:
            temp = self.head
            self.head = self.head.next
            self.head.prev = None
            temp.next = None
            self.nodes -= 1
            return temp
    
    def deleteMid(self,pos):
        if self.head is None:
            return None
        else:
            if pos == 1:
                self.deleteStart()
            elif pos >= self.nodes:
                self.deleteEnd()
            else:
                curr, temp = self.head, self.head.next
                for _ in range(pos-2):
                    curr, temp = curr.next, temp.next
                curr.next = temp.next
                temp.next.prev = curr
                temp.prev = None
                temp.next = None
                self.nodes -= 1
                return temp
    
    def deleteEnd(self):
        if self.head is None:
            return None
        else:
            temp = self.tail
            self.tail = self.tail.prev
            temp.prev = None
            self.tail.next = None
            self.nodes -= 1
            return temp
    
    def display(self):
        if self.head is None:
            print('Empty Doubly Linked List')
        else:
            temp = self.head
            while temp is not None:
                print(temp.data, end='<-->')
                temp = temp.next
            print()
    
    def reverseDisplay(self):
        if self.head is None:
            print('Empty Doubly Linked List')
        else:
            temp = self.tail
            while temp is not None:
                print(temp.data, end='<-->')
                temp = temp.prev
            print()

File: JavaScript/Doubly Linked List.js
class Node {
    constructor(data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

class DoublyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.nodes = 0;
    }
    insertStart(data) {
        const newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
        }
        else {
            newNode.next = this.head;
            this.head.prev = newNode;
            this.head = newNode;
        }
        this.nodes += 1;
    }
    insertMid(pos, data) {
        if (this.head == null) {
            const newNode = new Node(data);
            this.head = newNode;
            this.tail = newNode;
            this.nodes += 1;
        }
        else {
            if (pos == 1)
                this.insertStart(data);
            else if (pos >= this.nodes)
                this.insertEnd(data);
            else {
                const newNode = new Node(data);
                let curr = this.head, temp = this.head.next;
                for (let i=0; i<pos-2; i++) {
                    curr = curr.next;
                    temp = temp.next;
                }
                curr.next = newNode;
                newNode.prev = curr;
                newNode.next = temp;
                temp.prev = newNode;
                this.nodes += 1;
            }
        }
    }
    insertEnd(data) {
        const newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
        }
        else {
            this.tail.next = newNode;
            newNode.prev = this.tail;
            this.tail = newNode;
        }
        this.nodes += 1;
    }
    deleteStart() {
        if (this.head == null)
            console.log('Empty Linked List');
        else {
            const temp = this.head;
            this.head = this.head.next;
            this.head.prev = null;
            temp.next = null;
            this.nodes -= 1;
        }
    }
    deleteMid(pos) {
        if (this.head == null)
            console.log('Empty Linked List');
        else {
            if (pos == 1)
                this.deleteStart();
            else if (pos >= this.nodes)
                this.deleteEnd();
            else {
                let curr = this.head, temp = this.head.next;
                for (let i=0; i<pos-2; i++) {
                    curr = curr.next;
                    temp = temp.next;
                }
                curr.next = temp.next;
                temp.next.prev = curr;
                temp.next = null;
                temp.prev = null;
                this.nodes -= 1;
            }
        }
    }
    deleteEnd() {
        if (this.head == null)
            console.log('Empty Linked List');
        else {
            const temp = this.tail;
            this.tail = this.tail.prev;
            temp.prev = null;
            this.tail.next = null;
            this.nodes -= 1;
        }
    }
    display() {
        if (this.head === null)
            console.log('Emppty Doubly Linked List');
        else {
            let temp = this.head;
            let data = ``;
            for (let i=0; i<this.nodes; i++) {
                data += `${temp.data}<-->`
                temp = temp.next;
            }
        }
        console.log(data);
    }
    reverseDisplay() {
        let temp = this.tail;
        let data = ``;
        for (let i=0; i<this.nodes; i++) {
            data += `${temp.data}<-->`
            temp = temp.prev;
        }
        console.log(data);
    }
}
